File: C/15_DP1/BOJ_09_10844_쉬운계단수.c
#include <stdio.h>
main() {
	int n;
	int dp[101][10];
	scanf("%d", &n);
	dp[1][0] = 0;
	for (int i = 1; i < 10; i++) {
		dp[1][i] = 1;
	}
	for (int i = 2; i <= n; i++) {
		for (int j = 0; j < 10; j++) {
			if (j == 0) {
				dp[i][0] = dp[i - 1][1] % 1000000000;
			}
			else if (j == 9) {
				dp[i][9] = dp[i - 1][8] % 1000000000;
			}
			else {
				dp[i][j] = (dp[i - 1][j - 1] + dp[i - 1][j + 1]) % 1000000000;
			}
		}
	}
	int result = 0;
	for (int i = 0; i < 10; i++) {
		result = (result + dp[n][i]) % 1000000000;
	}
	printf("%d", result % 1000000000);
}

File: JAVA/15_DP1/BOJ_09_10844_쉬운계단수.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {

    static Integer[][] dp;
    static int n;

    public static int stair(int cursor, int value) {
        if (cursor == n) {
            return dp[cursor][value];
        }
        if (dp[cursor][value] == null) {
            if (value == 0) {
                dp[cursor][value] = stair(cursor + 1, 1);
            } else if (value == 9) {
                dp[cursor][value] = stair(cursor + 1, 8);
            } else {
                dp[cursor][value] = stair(cursor + 1, value - 1) + stair(cursor + 1, value + 1);
            }
        }
        return dp[cursor][value] % 1000000000;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        dp = new Integer[n + 1][10];
        for (int i = 0; i < 10; i++) {
            dp[n][i] = 1;
        }
        int result = 0;
        for (int i = 1; i <= 9; i++) {
            result = (result + stair(1, i)) % 1000000000;
        }
        System.out.println(result % 1000000000);
    }
}